### ISSUE TITLE
Crash on Android:26

### ISSUE THREAD

**Issue by RichardPointon** [ok2c/httpclient-android-ext#1]
I am using the ext in an Android application that supports API 26 and seeing a crash no such field error when attempting to create a connection.

Caused by: java.lang.NoSuchFieldError: No static field INSTANCE of type Lorg/apache/http/message/BasicLineFormatter; in class Lorg/apache/http/message/BasicLineFormatter; or its superclasses (declaration of 'org.apache.http.message.BasicLineFormatter' appears in /system/framework/org.apache.http.legacy.boot.jar)

It appears that the legacy apache httpcore class is being loaded instead of the required httpcore 4.4 class that is linked by httpclient 4.5 classes.

The app works fine on API 28 or later as they do not ship with the legacy apache libraries in the class path. I do not have the support legacy apache setting enabled in the gradle file.

Sample application that shows the issue attached.
[MyApplication.zip](https://github.com/ok2c/httpclient-android-ext/files/4746321/MyApplication.zip)

Full stack trace:

2020-06-02 15:18:22.928 22682-23604/? E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #4
    Process: com.ivanti.android.filedirector, PID: 22682
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:353)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.NoSuchFieldError: No static field INSTANCE of type Lorg/apache/http/message/BasicLineFormatter; in class Lorg/apache/http/message/BasicLineFormatter; or its superclasses (declaration of 'org.apache.http.message.BasicLineFormatter' appears in /system/framework/org.apache.http.legacy.boot.jar)
        at org.apache.http.impl.io.DefaultHttpRequestWriterFactory.<init>(DefaultHttpRequestWriterFactory.java:53)
        at org.apache.http.impl.io.DefaultHttpRequestWriterFactory.<init>(DefaultHttpRequestWriterFactory.java:57)
        at org.apache.http.impl.io.DefaultHttpRequestWriterFactory.<clinit>(DefaultHttpRequestWriterFactory.java:47)
        at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<init>(ManagedHttpClientConnectionFactory.java:83)
        at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<init>(ManagedHttpClientConnectionFactory.java:96)
        at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<init>(ManagedHttpClientConnectionFactory.java:105)
        at org.apache.http.impl.conn.ManagedHttpClientConnectionFactory.<clinit>(ManagedHttpClientConnectionFactory.java:63)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager$InternalConnectionFactory.<init>(PoolingHttpClientConnectionManager.java:620)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.<init>(PoolingHttpClientConnectionManager.java:181)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.<init>(PoolingHttpClientConnectionManager.java:165)
        at com.ok2c.hc.android.http.AndroidHttpClientConnectionManagerBuilder.build(AndroidHttpClientConnectionManagerBuilder.java:147)
        at com.appsense.android.datanow.utilities.HttpUtils.createStandardHttpClient(HttpUtils.java:78)
        at com.appsense.android.datanow.providers.datanow.api.MultiUploadTask.upload(MultiUploadTask.java:301)
        at com.appsense.android.datanow.providers.datanow.api.MultiUploadTask.doInBackground(MultiUploadTask.java:104)
        at com.appsense.android.datanow.providers.datanow.api.MultiUploadTask.doInBackground(MultiUploadTask.java:48)
        at android.os.AsyncTask$2.call(AsyncTask.java:333)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764)

**Comment by ok2c** [comment]
@RichardPointon I am not sure what I can do about it. I always thought that Google had stopped shipping their fork of Apache HttpClient as of Android version 26. I have no idea why `/system/framework/org.apache.http.legacy.boot.jar` still present on your application classpath.

Please consider using Apache HttpClient 5.0 extensions instead that should work with Android 19 upwards.

https://ok2c.github.io/httpclient-android-ext/hc5.html

**Comment by RichardPointon** [comment]
I was hoping you would have tried the sample application I attached which is a fresh Android Studio project that included your library. Would suggest you set the minimum SDK version to 28 as it doesn't work on 26.

**Comment by ok2c** [comment]
@RichardPointon I tested the library for compatibility with Android 26.

**Comment by RichardPointon** [comment]
Ok thanks for your help

**Comment by ok2c** [comment]
@RichardPointon I am working on a new patch release of the library upgrading it to Apache HttpClient 5.0.1. I will see if I can re-test the HttpClient 4.5 compatibility with SDK 26 but fundamentally I see little sense in putting a lot of efforts into HttpClient 4.5 branch given that HttpClient 5.0 i now GA.

**Comment by RichardPointon** [comment]
I didn't know 5 had gone GA so I will give that a try, thanks again

**Comment by ok2c** [comment]
@RichardPointon I upgraded Apache HttpClient 5.0 dependency to version 5.0.1 and pushed out new 0.1.1 release of Apache HttpClient 5.0.x extensions.

I also re-tested Apache HttpClient 4.5.x version and found out it no longer worked with Android 26. I am not sure what has changed since the release one year ago and am not going to invest any time finding it out. However I changed the documentation to state the minimal Android API level as 28 for Apache HttpClient 4.5.x extensions.
